### PR TITLE
ts/indicators-atr

### LINKED ISSUE #11840: Migrate `indicators/atr` to TypeScript

### PR DESCRIPTION
**Pull request by activist**
Closes #11840 , Migrated AtrIndicator to TypeScript.

### CODE REVIEW

File: ts/indicators/atr.src.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+ *
+ *  !!!!!!! SOURCE GETS TRANSPILED BY TYPESCRIPT. EDIT TS FILE ONLY. !!!!!!!
+ *
+ * */
+
+'use strict';
+
+import H from '../parts/Globals.js';
+
+declare global {
```
Comment (bre1470) [inline ts/indicators/atr.src.ts]: Missing doclet
```ts
/**
 * Internal types
 * @private
 */
```

File: ts/indicators/atr.src.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+var seriesType = H.seriesType,
+    UNDEFINED: undefined;
+
+/* eslint-disable valid-jsdoc */
+// Utils:
+/**
+ * @private
+ */
+function accumulateAverage(
+    points: Array<Array<(number|Array<number>)>>,
```
Comment (bre1470) [inline ts/indicators/atr.src.ts]: Maybe you can use the type `Array<[number|Array<number>]>` to narrow it down.

File: ts/indicators/atr.src.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                xValue = (xVal as any)[0],
+                yValue = yVal[0],
+                range = 1,
+                prevATR = 0,
+                TR = 0,
+                ATR: Array<Array<number>> = [],
+                xData = [],
+                yData = [],
+                point,
+                i,
```
Comment (bre1470) [inline ts/indicators/atr.src.ts]: Types missing.

File: ts/indicators/atr.src.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+ * is inherited from [chart.type](#chart.type).
+ *
+ * @extends   series,plotOptions.atr
+ * @since     6.0.0
+ * @product   highstock
+ * @excluding dataParser, dataURL
+ * @apioption series.atr
+ */
+
+''; // to include the above in jsdocs
```
Comment (bre1470) [inline ts/indicators/atr.src.ts]: Should be something like `to include the above in the js output`

File: ts/indicators/atr.src.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        TR = Math.max(HL, HCp, LCp);
+
+    return TR;
+}
+
+/**
+ * @private
+ */
+function populateAverage(
+    points: any,
```
Comment (bre1470) [inline ts/indicators/atr.src.ts]: This should be `Array<[number|Array<number>]>` too.

File: ts/indicators/atr.src.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                yValue = yVal[0],
+                range = 1,
+                prevATR = 0,
+                TR = 0,
+                ATR: Array<Array<number>> = [],
+                xData = [],
+                yData = [],
+                point,
+                i,
+                points: Array<Array<(number|Array<number>)>>;
```
Comment (bre1470) [inline ts/indicators/atr.src.ts]: This could be `Array<[number|Array<number>]>` as well.

File: ts/indicators/atr.src.ts (hunk 2 above)
Comment (activist) [reply ts/indicators/atr.src.ts]: Having a hard time to get this to work
```
Argument of type '[number, number[]]' is not assignable to parameter of type '[number | number[]]'.
  Types of property 'length' are incompatible.
    Type '2' is not assignable to type '1'.ts(2345)
``
Any clues?

File: ts/indicators/atr.src.ts (hunk 5 above)
Comment (activist) [reply ts/indicators/atr.src.ts]: Yes, it isn't used in the function so I tried the simple fix :)

File: ts/indicators/atr.src.ts (hunk 2 above)
Comment (bre1470) [reply ts/indicators/atr.src.ts]: You have to use the tuple everywhere in the module. You are mixing `[number , Array<number>]` (length = 2) with `Array<(number|Array<number>)>` (length = 1).

Comment (activist) [reply ts/indicators/atr.src.ts]: > Maybe you can use the type Array<[number|Array<number>]> to narrow it down.
Yes, I didn't get the tuple of number or Array<number> thing, but Array<[number,Array<number>]> makes sense :)

Comment (activist) [reply ts/indicators/atr.src.ts]: > Maybe you can use the type Array<[number|Array<number>]> to narrow it down.

Yes, I didn't get the whole tuple of number or Array<number> thing, but `[number, Array<number>] makes sense :)

**Comment by activist** [comment]
@bre1470, please review revised code with fixes. Slowly getting there.. :)

**Pull request merged**